app/src/main/java/com/example/fishkeeper/newcatch/NewCatchViewModel.kt

```diff
@@ -0,0 +1,202 @@
+        get() = _eventSubmit
+
+    //////////////////////////////////////////////////////////
+    // Species: required Stringfield
+    //////////////////////////////////////////////////////////
+    val species = MutableLiveData<String>()
+    private val _speciesError = MutableLiveData<Int?>()
+    val speciesError: LiveData<Int?>
+        get() = _speciesError
+
+    //////////////////////////////////////////////////////////
+    // Lure/fly type: required String field
+    //////////////////////////////////////////////////////////
+    val lure = MutableLiveData<String>()
+    private val _lureError = MutableLiveData<Int?>()
+    val lureError: LiveData<Int?>
+        get() = _lureError
+
+    //////////////////////////////////////////////////////////
+    // Hook size: required Int field
+    //////////////////////////////////////////////////////////
+    val hookSize = MutableLiveData<Int?>()
+    private val _hookSizeError = MutableLiveData<Int?>()
+    val hookSizeError: LiveData<Int?>
+        get() = _hookSizeError
+
+    //////////////////////////////////////////////////////////
+    // Length: optional Float field (String for now...)
+    //////////////////////////////////////////////////////////
+    val length = MutableLiveData<String>() //TODO make a float
+    private val _lengthError = MutableLiveData<Int?>()
+    val lengthError: LiveData<Int?>
+        get() = _lengthError
+
+    //////////////////////////////////////////////////////////
+    // Weight: optional Float field (String for now...)
+    //////////////////////////////////////////////////////////
+    val weight = MutableLiveData<String>() //TODO make a float
+    private val _weightError = MutableLiveData<Int?>()
+    val weightError: LiveData<Int?>
+        get() = _weightError
+
+
+    //////////////////////////////////////////////////////////
+    // Weight: optional Float field (String for now...)
+    //////////////////////////////////////////////////////////
+    private val _postComplete = MutableLiveData<Boolean?>()
+    val postComplete: LiveData<Boolean?>
+        get() = _postComplete
+
+    override fun onCleared() {
+        super.onCleared()
+        compositeDisposable.dispose()
+    }
+
+    fun eventSubmit() {
+        _eventSubmit.value = true
+    }
+
+    fun eventSubmitHandled() {
+        _eventSubmit.value = false
+    }
+
+    fun submitCatch() {
+        Log.d(TAG, "species: ${species.value}")
+        Log.d(TAG, "hook: ${hookSize.value}")
+        val isValid = validateData()
+        if (isValid) {
+            sendCatch()
+        }
+    }
+
+    private fun sendCatch() {
+
+        if (species.value == null || lure.value == null || hookSize.value == null) {
+            return
+        }
+
+        val newCatch = CatchPost(
+            null, // TODO add lat
+            null, // TODO add long
+            null, // TODO add altitude
+            length.value?.toDoubleOrNull(),
+            species.value!!,
```

Comment:
What does the double bang operator do here?

Reply:
`species.value` is of type "nullable String", meaning it could hold a `String`...or it could hold null.  The `CatchPost` constructor requires a `String` for the 5th parameter (the species).  I've did a null check up above (line 95), so here I'm using the `!!` operator to forcibly convert the nullable String to be of type String.  It's not good practice--I'm pretty sure there's some way to do the null check so that the compiler recognizes that they value can't be null on this line of code...but I'm not sure how, hence the `!!`.

Reply:
https://kotlinlang.org/docs/reference/null-safety.html for better explanation of null safety.